Add dilation convolution.

hzxie opened the pull request:
I added support for dilated convolution.

hzxie commented:
The implementation of dilation looks like the following image.
I think it is the same as the [article](https://towardsdatascience.com/understanding-2d-dilated-convolution-operation-with-examples-in-numpy-and-tensorflow-with-d376b3972b25).

![20180329203300892](https://user-images.githubusercontent.com/1730504/44190871-449c7400-a15b-11e8-89cc-fd128a377cd0.gif)

hzxie commented:
The dilated convolution is calculated as following.
![20180329203300892](https://user-images.githubusercontent.com/1730504/44191071-30a54200-a15c-11e8-9aac-5012e22ef2a9.gif)

I think the commited code is correct.
Let's consider the dilation convolution for the first row (without moving down).
The dilated convolution kernel starts at the position when left top corder of the kernel at (0, 0).
It ends at the position when right top corder of the kernel at (0, output_width).

Look back to the GIF image, the dilation convolution kernel starts at (0, 0) (left top corner of kernel) and ends at (0, 6) (right top corder of kernel). It moves 3 times.

Then, let's look back to the committed code.

```
for (size_t y = 0; y < oh; y++) {
  const float_t *pin_line = pin;
  for (size_t x = 0; x < ow; x++) {
    const float_t *pin_element = pin_line;
    const float_t *pw_element  = pw;
    float_t sum{0};
    // should be optimized for small kernel(3x3,5x5)
    for (size_t wy = 0; wy < kh; wy++) {    // NOLINT
      for (size_t wx = 0; wx < kw; wx++) {  // NOLINT
        sum += pw_element[wx] * pin_element[wx * w_dilation];
      }
      pw_element += kw;
      pin_element += iw * h_dilation;
    }
    pout[x] += sum;
    pin_line += elem_stride;
  }
  pout += ow;
  pin += line_stride;
}
```

The convolutional kernel moves `ow` times (`for (size_t x = 0; x < ow; x++)`).
And `ow` indicates `params.out.width_` which was updated at `conv_out_length()` function in `util/util.h`.

hzxie commented:
Here's the result calculated by PyTorch.
Suppose the tensor `x` is
```
array([[ 0,  1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12, 13],
       [14, 15, 16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25, 26, 27],
       [28, 29, 30, 31, 32, 33, 34],
       [35, 36, 37, 38, 39, 40, 41],
       [42, 43, 44, 45, 46, 47, 48]])
```

The kernel values in convolutional layer `layer` is

```
array([[[[10, 11, 12],
         [13, 14, 15],
         [16, 17, 18]]]])
```

The definition of `layer` is 
```
layer = torch.nn.Conv2d(in_channels =1, out_channels =1, kernel_size=3, stride=1, padding=0, dilation=2, bias=False)
```

The output is 

```
tensor([[[[2280., 2406., 2532.],
          [3162., 3288., 3414.],
          [4044., 4170., 4296.]]]], grad_fn=<ThnnConvDilated2DBackward>)
```

hzxie commented:
I tested my code with following stub

```
#include <cstdlib>
#include <iostream>
#include <vector>

#include "tiny_dnn/tiny_dnn.h"

class TestNet : public tiny_dnn::network<tiny_dnn::sequential> {
 public:
  explicit TestNet(const std::string &name = "")
    : tiny_dnn::network<tiny_dnn::sequential>(name) {
    using conv     = tiny_dnn::layers::conv;
    *this << conv(7, 7, 3, 3, 1, 1, tiny_dnn::padding::valid, true, 1, 1, 2, 2);
  }
};

int main() {
  TestNet nn;
  nn.weight_init(tiny_dnn::weight_init::gaussian(0));
  nn.bias_init(tiny_dnn::weight_init::constant(0));
  nn.init_weight();
  
  int init_value = 0;
  tiny_dnn::vec_t in(7 * 7 * 1);
  for ( auto itr = in.begin(); itr != in.end(); ++ itr ) {
    *itr = init_value ++;
  }

  std::cout << "Before predict" << std::endl;
  auto res = nn.predict(in);
  std::cout << "After predict" << std::endl;
  for ( auto itr = res.begin(); itr != res.end(); ++ itr ) {
    std::cout << *itr << std::endl;
  }
}
```

To initialize the weights in convolutional layer, I changed the `tiny_dnn::weight_init::gaussian(2)` as following

```
class gaussian : public scalable {
 public:
  gaussian() : scalable(float_t{1}) {}
  explicit gaussian(float_t sigma) : scalable(sigma) {}

  void fill(vec_t *weight, size_t fan_in, size_t fan_out) override {
    CNN_UNREFERENCED_PARAMETER(fan_in);
    CNN_UNREFERENCED_PARAMETER(fan_out);

    // gaussian_rand(weight->begin(), weight->end(), float_t{0}, scale_);
    std::cout << "FUCK ";
    int init_value = 10;
    for ( auto itr = weight->begin(); itr != weight->end(); ++ itr ) {
      *itr = init_value ++;
      std::cout << *itr << std::endl;
    }
    std::cout << "YOUR FATHER\n";
  }
};
```

And I got the same result as PyTorch:

```
2280
2406
2532
3162
3288
3414
4044
4170
4296
```

hzxie commented:
Private Note:

g++ main.cpp -o main -I/Users/hzxie/Desktop/tiny-dnn -std=c++14 -lpthread

Please ignore this message.

beru commented:
@hzxie 

> I don't know how to run this test.

IIRC, you can run the test in this way.

```sh
cmake -DBUILD_EXAMPLES=ON -DBUILD_TESTS ..
make
./tiny_dnn_test
```

beru commented:
You need to update this code.
https://github.com/hzxie/tiny-dnn/blob/459a05202f15e2707e159c74f4ccb112b16356f3/test/test_core.h#L105

It's mentioned in CI job's log.
https://ci.appveyor.com/project/tinydnn/tiny-dnn/build/master-1919/job/q5poa5yfnig3gkd2#L2711

beru commented:
You also need to use `clang-format` to pass travis-ci jobs.
Please check its logs.
https://travis-ci.org/tiny-dnn/tiny-dnn/builds/417240790?utm_source=github_status&utm_medium=notification

But I personally think doing so with every pull requets is too much for contributors.
@edgarriba Can't we stop having this rule? If the project isn't active, keeping strict rules is meaningless.

hzxie commented:
@beru 
I cannot build test with your command and got following error message:

```
Parse error in command line argument: -DBUILD_TESTS
Should be: VAR:type=value
CMake Error: No cmake script provided.
CMake Error: Problem processing arguments. Aborting.
```

edgarriba commented:
@beru we can remove it. But we what's the alternative to not get a messy and unstructured code ?

beru commented:
@hzxie 
> I cannot build test with your command and got following error message:

Sorry, I misspecified the instructions.

```sh
cmake -DBUILD_EXAMPLES=ON -DBUILD_TESTS=ON ..
make
./tiny_dnn_test
```

hzxie commented:
@beru @edgarriba 
The problem of `clang-format` has been solved.
But I got trouble in unit tests on serialization:

The `LoadAndConstruct` struct of convolution_layers is pasted as follows:

```
template <>
struct LoadAndConstruct<tiny_dnn::convolutional_layer> {
  template <class Archive>
  static void load_and_construct(
    Archive &ar, cereal::construct<tiny_dnn::convolutional_layer> &construct) {
    size_t w_width, w_height, out_ch, w_stride, h_stride, w_dilation,
      h_dilation;
    bool has_bias;
    tiny_dnn::shape3d in;
    tiny_dnn::padding pad_type;
    tiny_dnn::core::connection_table tbl;

    ::detail::arc(ar, ::detail::make_nvp("in_size", in),
                  ::detail::make_nvp("window_width", w_width),
                  ::detail::make_nvp("window_height", w_height),
                  ::detail::make_nvp("out_channels", out_ch),
                  ::detail::make_nvp("connection_table", tbl),
                  ::detail::make_nvp("pad_type", pad_type),
                  ::detail::make_nvp("has_bias", has_bias),
                  ::detail::make_nvp("w_stride", w_stride),
                  ::detail::make_nvp("h_stride", h_stride),
                  ::detail::make_nvp("w_dilation", w_dilation),
                  ::detail::make_nvp("h_dilation", h_dilation));

    construct(in.width_, in.height_, w_width, w_height, in.depth_, out_ch, tbl,
              pad_type, has_bias, w_stride, h_stride, w_dilation, h_dilation);
  }
};
```

Can you help me solve this problem?
Thank you.

beru commented:
@edgarriba 
> we can remove it. But we what's the alternative to not get a messy and unstructured code ?

Maintainers like us occasionally may apply `scripts/clang-format.sh` script.
My opinion is new contributors shouldn't be bothered with it.

hzxie commented:
@beru @edgarriba 
Another problem, my newly created unit testing case has the following error message:

```
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
```
See: https://ci.appveyor.com/project/tinydnn/tiny-dnn/build/master-1927/job/28ee9ahr6yo0buer#L2947

The unit testing code can be found in: https://github.com/tiny-dnn/tiny-dnn/pull/978/files#diff-bb1b7f8be34ab56fb6f33c41089b4bfdR207

beru commented:
@hzxie This string should be updated.

https://github.com/hzxie/tiny-dnn/blob/38e37f368a157efabde6e150d9e50bd4259ac90d/test/test_serialization.h#L174

I guess.. each of the members have to be there for deserialization with cereal library.

hzxie commented:
@beru
Yes. I have updated it. But haven't pushed.
But there're other failed unit testing cases:

- network.read_write
- serialization. sequential_to_json

How to solve those two cases?

hzxie commented:
@beru 

I updated line 174 in `test_serialization.h`.
But still failing and got following message:

```
[ RUN      ] serialization.serialize_conv
unknown file: Failure
C++ exception with description "Failed to read 4 bytes from input stream! Read 0" thrown in the test body.
```

hzxie commented:
@beru @edgarriba 
All CI errors have been solved.

hzxie commented:
<img width="920" alt="screen shot 2018-08-21 at 12 03 54 am" src="https://user-images.githubusercontent.com/1730504/44352141-1f4e9380-a4d6-11e8-9115-a4db3f376c4e.png">
I can pass all tests on my Mac.
But in Travis CI, it has several errors.
Can you tell me the reason?

beru commented:
@hzxie 
> I can pass all tests on my Mac. But in Travis CI, it has several errors. Can you tell me the reason?

I don't have Mac right now so let me guess.

https://travis-ci.org/tiny-dnn/tiny-dnn/jobs/418293807#L1216

https://github.com/tiny-dnn/tiny-dnn/blob/9fed09edb401b17467b75b35cb6feaa87dd54fe6/test/test_convolutional_layer.h#L222

Address sanitizer detected buffer boundary violation in the test, `TEST(convolutional, with_dilation)`.
I suspect your implementation does not handle corner cases (literally...) properly.

edgarriba commented:
@beru @hzxie apart from that, osx installation crashes now in Travis. Couldn't find why yet

hzxie commented:
@beru @edgarriba 
I changed my code to the original:
```
// should be optimized for small kernel(3x3,5x5)
for (size_t wy = 0; wy < kh; wy++) {    // NOLINT
 for (size_t wx = 0; wx < kw; wx++) {  // NOLINT
   sum += pw_element[wx] * pin_element[wx];
   // sum += pw_element[wx] * pin_element[wx * w_dilation];
 }
 pw_element += kw;
 pin_element += iw;
 // pin_element += iw * h_dilation;
}
```
But the unit testing still leads to Segmentation fault.

I also added print statement to check the position where the error occurred:

```
inline void conv2d_op_internal(const tensor_t &in_data,
                               const vec_t &W,
                               const vec_t &bias,
                               tensor_t &out_data,
                               const core::conv_params &params,
                               const bool parallelize) {
  std::cout << "In conv2d_op_internal function" << std::endl;
  for_(parallelize, 0u, in_data.size(),
```

```
if (engine == core::backend_t::internal) {
  std::cout << "In compute function of Conv2dOp" << std::endl;
  kernels::conv2d_op_internal(in_data, W[0], bias[0], out_data, params,
                              context.parallelize());
} else if (engine == core::backend_t::nnpack) {
```

And I got following output:

```
In compute function of Conv2dOp
```

The first statement in `conv2d_op_internal` has not been executed.
So strange!!

hzxie commented:
@beru 
I believe my implementation of dilated convolution is correct.

hzxie commented:
The unit test can sometimes be passed in Ubuntu and Mac. (Sometimes they will crashed)

![screenshot from 2018-08-21 15-43-26](https://user-images.githubusercontent.com/1730504/44388644-3388a400-a55b-11e8-862a-3ad4bfc0c864.png)

hzxie commented:
If I delete other unit testing cases in `test.cpp` as following:

```
// #include "test_activation_layer.h"
// #include "test_average_pooling_layer.h"
// #include "test_network.h"
// // TODO(yida): fix broken test
// // #include "test_average_unpooling_layer.h"
// #include "test_batch_norm_layer.h"
// #include "test_concat_layer.h"
#include "test_convolutional_layer.h"
// #include "test_core.h"
// #include "test_deconvolutional_layer.h"
// #include "test_dropout_layer.h"
// #include "test_fully_connected_layer.h"
// #include "test_global_average_pooling_layer.h"
// #include "test_integration.h"
// #include "test_large_thread_count.h"
// #include "test_lrn_layer.h"
// #include "test_max_pooling_layer.h"
// #include "test_models.h"
// #include "test_node.h"
// #include "test_nodes.h"
// #include "test_optimizers.h"
// #include "test_power_layer.h"
// #include "test_quantization.h"
// #include "test_quantized_convolutional_layer.h"
// #include "test_quantized_deconvolutional_layer.h"
// #include "test_slice_layer.h"
// #include "test_target_cost.h"
// #include "test_tensor.h"

// #include "test_gru_cell.h"
// #include "test_lstm_cell.h"
// #include "test_rnn_cell.h"
```
The unit testing cases for dilation convolution can be passed (always) in both Ubuntu and Mac.
![screenshot from 2018-08-21 16-13-51](https://user-images.githubusercontent.com/1730504/44389430-69c72300-a55d-11e8-970c-8db4d8519725.png)

So, what's the problem?

hzxie commented:
I recover to the original code (without dilation) as following:

```
for (size_t wy = 0; wy < kh; wy++) {    // NOLINT
 for (size_t wx = 0; wx < kw; wx++) {  // NOLINT
   sum += pw_element[wx] * pin_element[wx];
   // sum += pw_element[wx] * pin_element[wx * w_dilation];
 }
 pw_element += kw;
 pin_element += iw;
 // pin_element += iw * h_dilation;
}
```
But the CI is still failing. So I don't think there's something wrong with my  implementation of dilated convolution. Maybe there's something wrong with somewhere else (Maybe unit testing?).

See: https://travis-ci.org/tiny-dnn/tiny-dnn/jobs/418633230#L1330

hzxie commented:
@edgarriba 
No.
What does this option mean?

hzxie commented:
@beru @edgarriba 
I know what the problem is!!!
There's something wrong with `conv` *without* bias.

If you create a convolution layer without bias. The following line will cause illegal memory access.

```
const tensor_t &bias    = context.input(2);
```

hzxie commented:
@beru @edgarriba 
There's nothing wrong with my dilated convolution code but convolution layer without bias!!

hzxie commented:
@beru @edgarriba 
The CI on Linux has been passed.
See: https://travis-ci.org/tiny-dnn/tiny-dnn/builds/418673860

<img width="1392" alt="screen shot 2018-08-21 at 8 55 57 pm" src="https://user-images.githubusercontent.com/1730504/44402521-a22e2780-a584-11e8-88bd-65b9ac6f841a.png">

I'm working on fixing errors on macOS.
Please consider merging this PR.
Thank you for reviewing my code and answering my questions.

edgarriba commented on .travis/install.sh:
```diff
@@ -11,6 +11,9 @@ if [ "$TRAVIS_OS_NAME" == "osx" ]; then
     brew install pkg-config
   else
+    # Fix errors on Travis CI with macOS due to the conflict of python@2 
+    brew unlink python
```
nice !
